File: backend/app/api/tickers.py
"""Ticker existence validation (CR128).

GET /v1/tickers/validate?ticker=XXX

Called by the Flutter client at submit-time from all three ticker-entry
points (Convene the Room, Start Trade, Add to Watchlist) before the actual
action request. No auth — read-only reference lookup, no user-specific data,
same shape as `/v1/glossary/*`.
"""

from __future__ import annotations

from fastapi import APIRouter, Query

from app.db import get_session
from app.schemas.tickers import TickerSuggestion, TickerValidateResponse
from app.services.ticker_reference import lookup_ticker, suggest_closest

router = APIRouter(prefix="/v1/tickers", tags=["tickers"])


@router.get("/validate", response_model=TickerValidateResponse)
async def validate_ticker(ticker: str = Query(..., min_length=1)) -> TickerValidateResponse:
    normalized = ticker.upper().strip()
    with get_session() as session:
        row = lookup_ticker(session, normalized)
        if row is not None:
            return TickerValidateResponse(
                ticker=row.symbol,
                exists=True,
                company_name=row.company_name,
                exchange=row.exchange,
            )
        candidates = suggest_closest(session, normalized, limit=1)
        suggestion = None
        if candidates:
            match = candidates[0]
            suggestion = TickerSuggestion(
                ticker=match.symbol,
                company_name=match.company_name,
                exchange=match.exchange,
            )
        return TickerValidateResponse(ticker=normalized, exists=False, suggestion=suggestion)

File: mobile/lib/models/tickers.dart
/// Ticker existence validation — client-side models (CR128).
/// Mirrors backend/app/schemas/tickers.py.
library;

class TickerSuggestion {
  const TickerSuggestion({
    required this.ticker,
    required this.companyName,
    required this.exchange,
  });

  final String ticker;
  final String companyName;
  final String exchange;

  factory TickerSuggestion.fromJson(Map<String, dynamic> j) {
    return TickerSuggestion(
      ticker: j['ticker'] as String,
      companyName: j['company_name'] as String,
      exchange: j['exchange'] as String,
    );
  }
}

class TickerValidation {
  const TickerValidation({
    required this.ticker,
    required this.exists,
    this.companyName,
    this.exchange,
    this.suggestion,
  });

  final String ticker;
  final bool exists;
  final String? companyName;
  final String? exchange;
  final TickerSuggestion? suggestion;

  factory TickerValidation.fromJson(Map<String, dynamic> j) {
    return TickerValidation(
      ticker: j['ticker'] as String,
      exists: j['exists'] as bool,
      companyName: j['company_name'] as String?,
      exchange: j['exchange'] as String?,
      suggestion: j['suggestion'] == null
          ? null
          : TickerSuggestion.fromJson(j['suggestion'] as Map<String, dynamic>),
    );
  }
}
